lib/plugins/cleaner.rb

```diff
@@ -0,0 +1,47 @@
+require 'rubygems/package'
+require_relative '../moonshot/creds_helper'
+
+module Moonshot
+  module Plugins
+    # Moonshot plugin class for removing artifacts when a stage gets deleted.
+    class Cleaner # rubocop:disable Metrics/ClassLength
+      include Moonshot::CredsHelper
+
+      def initialize(bucket_name)
+        @bucket_name = bucket_name
+      end
+
+      def pre_delete(resources)
+        app_name = resources.stack.app_name
+        stack_name = resources.stack.name
+        resources.ilog.start "Starting to delete artifacts for app: #{app_name} and stack: #{stack_name}" do |s|
+          arts = retrieve_artifact_names(stack_name)
+          delete_artifacts arts
+          s.success "Deleted app: '#{app_name}' and stack: '#{stack_name}' bucket: '#{@bucket_name}'"
+        end
+      end
+
+      private
+
+      def retrieve_artifact_names(name)
+        objs = s3_client.list_objects(bucket: @bucket_name, prefix: "#{name}")
+        valid_keys = []
+        objs.contents.each { |o| valid_keys << o.key if o.key =~ /^#{name}[\.-]/ }
+        raise Thor::Error,
+              "Artifact not found with name:" \
+              "#{name} in bucket: #{@bucket_name}!" if valid_keys.count < 1
+        valid_keys
+      end
+
+      def delete_artifacts(artifacts)
+        objects = artifacts.map { |a| { key: a } }
+        resp = s3_client.delete_objects(bucket: @bucket_name,
+                                        delete: { objects: objects,
+                                                  quiet: false } )
+        raise Thor::Error,
```

Comment:
Being that this is currently a pre_delete step, I wonder if throwing an error here is what everyone would expect. When it comes down to cost, terminating the stack is significantly more important than the deployments in the bucket. Perhaps this is another great reason why we want this as a post_delete hook. Also, be advised that callers MAY (us included) have additional hooks running and by throwing an exception here, you kill any following hooks which might not be entirely expected.

I don't think we've really discussed how hooks should work when they failed to do their job. Should they raise an exception and kill the whole delete chain or should they just log an error message and continue? If they do log an error message, how does that affect the exit code of the Moonshot::CLI. I bet if we just log an error message line it would exit 0 which is obviously wrong still so there is clearly some work to be done here in general.